osu.Game.Rulesets.Taiko/UI/DrumTouchInputArea.cs

```diff
@@ -0,0 +1,141 @@
+        private TaikoAction mouseAction;
+
+        // A map of (Finger Index OnTouchDown -> Which Taiko action was pressed), so that the corresponding action can be released OnTouchUp is released even if the touch position moved
+        private Dictionary<TouchSource, TaikoAction> touchActions = new Dictionary<TouchSource, TaikoAction>(Enum.GetNames(typeof(TouchSource)).Length);
+
+        private Playfield playfield;
+
+        public DrumTouchInputArea(Playfield playfield) {
+            this.playfield = playfield;
+
+            RelativeSizeAxes = Axes.Both;
+            RelativePositionAxes = Axes.Both;
+            Children = new Drawable[]
+            {
+                new Box() {
+                    Alpha = 0.0f,
+                    Colour = new OsuColour().Blue,
+
+                    RelativeSizeAxes = Axes.Both,
+                    RelativePositionAxes = Axes.Both,
+                },
+                new Container() {
+                    RelativeSizeAxes = Axes.Both,
+                    RelativePositionAxes = Axes.Both,
+
+                    Anchor = Anchor.BottomCentre,
+                    Origin = Anchor.BottomCentre,
+
+                    Children = new Drawable[]
+                    {
+                        touchInputDrum = new InputDrum(playfield.HitObjectContainer) {
+                            Anchor = Anchor.Centre,
+                            Origin = Anchor.Centre,
+                        },
+                    }
+                },
+            };
+
+        }
+        protected override void LoadComplete()
+        {
+            keyBindingContainer = taikoInputManager?.KeyBindingContainer;
+
+            Padding = new MarginPadding {
+                Top =  playfield.ScreenSpaceDrawQuad.BottomLeft.Y,
+                Bottom = -DrawHeight * overhangPercent,
+            };
+        }
+
+        protected override bool OnMouseDown(MouseDownEvent e)
+        {
+            mouseAction = getTaikoActionFromInput(e.ScreenSpaceMouseDownPosition);
+            keyBindingContainer?.TriggerPressed(mouseAction);
+            return base.OnMouseDown(e);
+        }
+
+        protected override void OnMouseUp(MouseUpEvent e)
+        {
+            keyBindingContainer?.TriggerReleased(mouseAction);
+            base.OnMouseUp(e);
+        }
+
+        protected override bool OnTouchDown(TouchDownEvent e)
+        {
+            TaikoAction taikoAction = getTaikoActionFromInput(e.ScreenSpaceTouchDownPosition);
+            if (touchActions.ContainsKey(e.Touch.Source)) {
+                touchActions[e.Touch.Source] = taikoAction;
+            }
+            else {
+                touchActions.Add(e.Touch.Source, taikoAction);
+            }
+            keyBindingContainer?.TriggerPressed(touchActions[e.Touch.Source]);
+
+            return base.OnTouchDown(e);
```

Comment:
Pretty sure you actually want to be handling these events, else they will also appear as a mouse event. Have you tested this PR on a touch device?

Reply:
I'm a bit confused  - I think they are handled? Are you saying this will result in a double-input when tapping the screen (one from the mouse handling, and a duplicate from touch handling)?

I haven't tested on a touch device yet.

Reply:
Changed to `return true;` as of 317869078f470889330702b7b99b6146569b4f75